ml/train/losses/losses.go

```diff
@@ -398,3 +402,32 @@ func MakeHuberLoss(delta float64) LossFn {
 		return loss
 	}
 }
+
+// CosineSimilarityLoss computes the cosine similarity loss between two predictions.
+//
+// It assumes two predictions are provided which should be tensors of rank 2 with the same dimensions (N, D).
+// It also assumes there is one label tensor with rank 2 with dimensions (N, 1) that contains the target cosine similarity.
+// The loss will then calculate the residuals between the cosine similarity of the two prediction tensors row-wise
+// and the target cosine similarity, and return the mean of the squared residuals.
+func CosineSimilarityLoss(labels, predictions []*context.Node) *context.Node {
+	predictionsLeft := predictions[0]
+	predictionsRight := predictions[1]
+	scores := labels[0]
+
+	if predictionsLeft.Shape().Rank() != 2 {
+		panic(fmt.Errorf("expected rank 2, got %d", predictionsLeft.Shape().Rank()))
+	}
+	if predictionsRight.Shape().Rank() != 2 {
+		panic(fmt.Errorf("expected rank 2, got %d", predictionsLeft.Shape().Rank()))
+	}
+	if scores.Shape().Rank() != 2 {
+		panic(fmt.Errorf("expected rank 2, got %d", predictionsLeft.Shape().Rank()))
+	}
+	dotProduct := graph.ReduceAndKeep(graph.Mul(predictionsLeft, predictionsRight), graph.ReduceSum, 1)
```

Comment:
no need for the `graph.` prefix, since `graph` is imported as "." (in the current scope).

In files that do numeric computations for GoMLX, I do this as a convention. For other files that are most related other things I don't.

Btw, optional, but maybe it will be more optimized (because the Einsum is optimized to do these multiplication/accumulation ... although it's possible that the way you wrote will be fused and optimized as well):

dotProduct := Einsum("ij,ij->i", predictionsLeft, predictionsRight)

A good reminder I should create a DotProduct function ...